core/src/services/hdfs_native/reader.rs

```diff
@@ -18,20 +18,40 @@
 use hdfs_native::file::FileReader;
 
 use crate::raw::*;
+use crate::services::hdfs_native::error::parse_hdfs_error;
 use crate::*;
 
 pub struct HdfsNativeReader {
-    _f: FileReader,
+    f: FileReader,
 }
 
 impl HdfsNativeReader {
     pub fn new(f: FileReader) -> Self {
-        HdfsNativeReader { _f: f }
+        HdfsNativeReader { f }
     }
 }
 
 impl oio::Read for HdfsNativeReader {
-    async fn read_at(&self, _offset: u64, _limit: usize) -> Result<Buffer> {
-        todo!()
+    async fn read_at(&self, offset: u64, limit: usize) -> Result<Buffer> {
+        // Check for offset being too large for usize on 32-bit systems
+        if offset > usize::MAX as u64 {
```

Comment:
Please don't run checks in services.

Reply:
Where can we put this check ?

Reply:
> Where can we put this check ?

We don't need this check for current. It's more like an upstream issue that can't range from `i32::MAX..i64::MAX`. You can create an opendal issue to track this.

Reply:
Sure, created issue #4506